src/software/proto/message_translation/proto_creator_primitive_visitor.cpp

```diff
@@ -0,0 +1,159 @@
+#include "software/proto/message_translation/proto_creator_primitive_visitor.h"
+
+#include "shared/constants.h"
+#include "software/primitive/all_primitives.h"
+
+PrimitiveMsg ProtoCreatorPrimitiveVisitor::getProto()
+{
+    // If we've never visited a primitive (and so have never populated the
+    // `prim`) then throw an exception
+    if (!prim)
+    {
+        std::string err_msg = std::string(typeid(this).name()) + ": " + __func__ +
+                              " called without ever having visited anything";
+        throw std::runtime_error(err_msg);
+    }
+
+    return *prim;
```

Comment:
can we pass in the primitive to `getProto` and avoid this check?

Reply:
So the `visit` functions need to remain public so that the primitives can visit them, but made `getProto` private and exposed `PrimitiveMsg createPrimitiveMsg(const Primitive& primitive)`.

Great suggestion!

Reply:
I made a note to apply this to other visitors later too :+1: